[spec] Add `on_click` parameter for `st.link_button`

lukasmasuch opened the pull request:
Rendered spec: https://issues.streamlit.app/spec_renderer?pr=13197

## GitHub Issue Link (if applicable)

- Related feature request: https://github.com/streamlit/streamlit/issues/7453
- Related prototype: https://github.com/streamlit/streamlit/pull/13174

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

lukasmasuch committed:
[spec] Add `on_click` parameter for `st.link_button`

lukasmasuch committed:
Update

sfc-gh-lwilby commented on specs/0007-link-button-on-click/product-spec.md:
```diff
@@ -0,0 +1,114 @@
+- Current behavior—clicking opens the URL in a new tab
+- No app rerun occurs
+
+**`on_click="rerun"`:**
```
[question] Is there a use-case for this, or is it primarily for consistency with the other usages?

lukasmasuch replied:
In general or specific to `on_click="rerun"`? With `on_click="rerun"` it would just act as a normal button, returning true if clicked. 
Overall, we have a feature request for reacting to `link_button` clicks here: https://github.com/streamlit/streamlit/issues/7453 with a couple of upvotes. I think there are some niche use-cases for it, but its also nice for consistency across our buttons (e.g. with `download_button`).

sfc-gh-lwilby replied:
Yes, just wondering about the "rerun" option, thanks!

sfc-gh-lwilby commented on specs/0007-link-button-on-click/product-spec.md:
```diff
@@ -0,0 +1,114 @@
+
+- Clicking opens the URL in a new tab AND triggers a rerun
+- The callback executes before the rest of the app runs
+- `args` and `kwargs` are passed to the callback
```
Wondering if we should provide some default args/kwargs, i.e. label, url and use these params to allow the user to provide additional params to the callback?

lukasmasuch replied:
You mean some args/kwargs that always get passed into the callback function? I'm on-board with revamping our callback system to allow that at some point, also see: [Optional, named parameters for callbacks](https://www.notion.so/snowflake-corp/Optional-named-parameters-for-callbacks-aaf026126d0c49dbb5fd75004be89854) and [Deprecate args/kwargs](https://www.notion.so/snowflake-corp/Deprecate-args-kwargs-e934ff2883af406bb1ee068783b4ec02). Its probably a bit of a bigger decision we need to do across all the callbacks.

sfc-gh-lwilby replied:
Yes, I think it is a common pattern that when you add a callback there would be some expected params that will be passed into the function (e.g. onClick and event). 

Maybe this is something we could allow users to control by the arguments they specify on the function itself, although, maybe that is more difficult to understand than the current pattern. 

It makes sense that we follow the existing pattern here and then have a separate discussion about a revamp!

sfc-gh-tteixeira commented:
Problem: ✅ 
Solution: ✅ 
Implementation: 🙈

jrieke reviewed (approved):
LGTM! Should probably add `key` for consistency just like in the Copilot suggestions (or is there a reason to leave it out here?). 

One tangential thought: If we implement #7464 (opening a link in the same tab), how would this work? Would it still trigger a rerun in the current app and then open the link (probably awkward), or ignore the rerun and open the link right away? I don't think this is a blocker, and we can probably think about it when we get to that issue, just wanted to mention it :)

lukasmasuch committed:
Update

lukasmasuch committed:
Update

lukasmasuch committed:
Merge remote-tracking branch 'upstream/develop' into spec/link-button-on-click

lukasmasuch committed:
Update

lukasmasuch commented on specs/0000-link-button-on-click/product-spec.md:
```diff
@@ -0,0 +1,114 @@
+    on_click: Literal["rerun", "ignore"] | Callable = "ignore",  # NEW
+    args: WidgetArgs | None = None,  # NEW
+    kwargs: WidgetKwargs | None = None,  # NEW
+)
```
Updated 👍

lukasmasuch commented on specs/0000-link-button-on-click/product-spec.md:
```diff
@@ -0,0 +1,114 @@
+|-----------|------|---------|-------------|
+| `on_click` | `"ignore"`, `"rerun"`, or `callable` | `"ignore"` | How the button should respond to clicks. `"ignore"` (default): No rerun, link opens in new tab only. `"rerun"`: Clicking triggers a rerun in addition to opening the link. `callable`: Clicking triggers a rerun and executes the callback before the rest of the app. |
+| `args` | `list \| tuple \| None` | `None` | Arguments to pass to the callback. |
+| `kwargs` | `dict \| None` | `None` | Keyword arguments to pass to the callback. |
```
Updated 👍

lukasmasuch commented on specs/0000-link-button-on-click/product-spec.md:
````diff
@@ -0,0 +1,114 @@
+```python
+st.link_button(
+    ...,
+    on_click: Literal["rerun", "ignore"] | Callable = "ignore",  # NEW
````
Updated 👍

lukasmasuch commented:
> Should probably add key for consistency

Yep, added it 👍 

> One tangential thought: If we implement https://github.com/streamlit/streamlit/issues/7464 (opening a link in the same tab), how would this work?

Added it to future considerations. I think this is something to figure out when we do open in the same window... but might be something to just prevent with an exception.

lukasmasuch committed:
Update
